tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/Services/WorkTokens/CosmosAsyncLockProvider.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.Azure.Cosmos;
+
+namespace Azure.Sdk.Tools.PipelineWitness.Services.WorkTokens
+{
+    public class CosmosAsyncLockProvider : IAsyncLockProvider
+    {
+        private readonly Container container;
+
+        public CosmosAsyncLockProvider(CosmosClient cosmosClient, string databaseName, string containerName)
+        {
+            if (cosmosClient == null)
+            {
+                throw new ArgumentNullException(nameof(cosmosClient));
+            }
+
+            this.container = cosmosClient.GetContainer(databaseName, containerName);
+        }
+
+        public async Task<IAsyncLock> GetLockAsync(string id, TimeSpan duration, CancellationToken cancellationToken)
+        {
+            var partitionKey = new PartitionKey(id);
```

Comment:
Would it be better to have a static partition key for locks like `LockProvider`? That way you would have `LockProvider/UpdateBuildDefinitions` vs. `UpdateBuildDefinitions/UpdateBuildDefinitions`. Unless you want a separate partition per lock for a specific reason?

Reply:
I'm not very familiar with how partition keys work though and what the perf impact of one is.

Reply:
partition keys don't establish a hierarchy.  They're used in cluster's hashing function for mapping documents to the physical shard containing the document.

Reply:
So does that mean we would have a shard per lock type?

Reply:
Each lock may move into a shard without constraint for other locks being in the same shard.  Because of the low document count and size, we would only have 1 physical partition with each document being in a unique logical partition.  The logical partitions have no effect on current perf, just the ability to shard in the future.